tiledb/sm/tile/chunked_buffer.h

```diff
@@ -0,0 +1,382 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines class ChunkedBuffer.
+ *
+ * The ChunkedBuffer class is used to represent a logically contigious buffer
+ * as a vector of individual buffers. These individual buffers are referred to
+ * as "chunk buffers". Each chunk buffer may be allocated individually, which
+ * will save memory in scenarios where the logically contigious buffer is
+ * sparsley allocated.
+ *
+ * After construction, the class must be initialized before performing IO. The
+ * initialization determines the following, independent usage paradigms:
+ *
+ * #1: Chunk Sizes: Fixed/Variable
+ * The chunk sizes must be either fixed or variable. An instance with fixed
+ * chunk sizes ensures that all chunk buffers are of equal size. The size of the
+ * last chunk buffer may be equal-to or less-than the other chunk sizes.
+ * Instances with fixed size chunks have a smaller memory footprint and have a
+ * smaller algorithmic complexity when performing IO. For variable sized chunks,
+ * each chunk size is independent from the others.
+ *
+ * #2: Chunk Buffer Addressing: Discrete/Contigious
+ * The addresses of the individual chunk buffers may or may not be virtually
+ * contigious. For example, the chunk addresses within a virtually contigious
+ * instance may be allocated at address 1024 and 1028, where the first chunk is
+ * of size 4. Non-contigious chunks (referred to as "discrete") may be allocated
+ * at any address. The trade-off is that the memory of each discrete chunk is
+ * managed individually, where contigious chunk buffers can be managed by the
+ * first chunk alone.
+ *
+ * #3: Memory Management: Internal/External
+ * The chunk buffers may be allocated and freed internally or externally.
+ * Internal memory management is exposed through the alloc_*() and free_*()
+ * routines. External memory management is exposed through the set_*() routines.
+ * Currently, this only supports external memory management for contigiously
+ * addressed buffers and internal memory management for discretely addressed
+ * buffers.
+ *
+ * Note that the ChunkedBuffer class does NOT support any concept of ownership.
+ * It is up to the caller to free the instance before destruction.
+ */
+
+#ifndef TILEDB_chunked_buffer_H
+#define TILEDB_chunked_buffer_H
+
+#include "tiledb/sm/misc/logger.h"
+#include "tiledb/sm/misc/status.h"
+
+#include <cinttypes>
+#include <vector>
+
+namespace tiledb {
+namespace sm {
+
+class ChunkedBuffer {
+ public:
+  enum BufferAddressing { CONTIGIOUS, DISCRETE };
```

Comment:
Typo: `CONTIGIOUS` -> `CONTIGUOUS`

Reply:
Good catch, I'll do a global find-and-replace to typos on 'CONTIGUOUS'.

Reply:
Done